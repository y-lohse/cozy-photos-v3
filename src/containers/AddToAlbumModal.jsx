import styles from '../styles/addToAlbum'

import React, { Component } from 'react'
import { connect } from 'react-redux'
import { translate } from 'cozy-ui/react/I18n'
import Modal, { ModalSection } from 'cozy-ui/react/Modal'
import classNames from 'classnames'

import Alerter from '../components/Alerter'
import CreateAlbumForm from '../components/CreateAlbumForm'
import SelectAlbumsForm from '../components/SelectAlbumsForm'
import Loading from '../components/Loading'

import { getSelectedIds } from '../ducks/selection'

import {
  fetchAlbums,
  getAlbumsList,
  createAlbum,
  addToAlbum,
  checkUniquenessOfAlbumName,
  cancelAddToAlbum,
  closeAddToAlbum
} from '../ducks/albums'

import { refetchSomePhotos } from '../ducks/timeline'

export class AddToAlbumModal extends Component {
  componentWillMount () {
    this.props.fetchAlbums()
  }

  render (props) {
    const {
      t,
      photos,
      albums,
      onDismiss,
      onSubmitNewAlbum,
      onSubmitSelectedAlbum
    } = props

    const fetchStatus = albums ? albums.fetchStatus : 'loading'
    const isFetchingAlbums = fetchStatus === 'pending' || fetchStatus === 'loading'

    return (
      <Modal
        title={t('Albums.add_photos.title')}
        secondaryAction={() => onDismiss()}
        >
        <ModalSection className={styles['coz-modal-section']}>
          <div className={classNames(styles['coz-create-album'])}>
            <CreateAlbumForm
              onSubmitNewAlbum={name => onSubmitNewAlbum(name, photos)}
              />
          </div>
          {
            isFetchingAlbums
            ? <Loading loadingType='albums_fetching' />
            : albums && albums.entries.length > 0
            ? <div className={classNames(styles['coz-select-album'])}>
              <SelectAlbumsForm albums={albums} onSubmitSelectedAlbum={album => onSubmitSelectedAlbum(album, photos)} />
            </div>
            : null
          }
        </ModalSection>
      </Modal>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    photos: getSelectedIds(state),
    albums: getAlbumsList(state)
  }
}

const mapDispatchToProps = (dispatch, ownProps) => ({
  onDismiss: () => {
    dispatch(cancelAddToAlbum())
  },
  onSubmitNewAlbum: async (name, photos) => {
    try {
      if (!name) {
        Alerter.error('Albums.create.error.name_missing')
        return
      }
      // TODO: we should not have to dispatch this method (see comments in redux-cozy-api)
      const unique = await dispatch(checkUniquenessOfAlbumName(name))
      if (!unique) {
        Alerter.error('Albums.create.error.already_exists', { name })
        return
      }
      const album = await dispatch(createAlbum(name, photos))
      // TODO: sadly we need to refetch the photos so that their relationships
      // property get updated and so that isRelated works when deleting a photo
      // that has just been added to an album
      await dispatch(refetchSomePhotos(photos))
      dispatch(closeAddToAlbum())
      Alerter.success('Albums.create.success', {name: album.name, smart_count: photos.length})
    } catch (error) {
      Alerter.error('Albums.create.error.generic')
    }
  },
  onSubmitSelectedAlbum: async (album, photos) => {
    try {
      const addedPhotos = await dispatch(addToAlbum(album, photos))
      if (addedPhotos.length !== photos.length) {
        Alerter.info('Alerter.photos.already_added_photo')
      } else {
        Alerter.success('Albums.add_photos.success', {name: album.name, smart_count: photos.length})
      }
      // TODO: sadly we need to refetch the photos so that their relationships
      // property get updated and so that isRelated works when deleting a photo
      // that has just been added to an album
      await dispatch(refetchSomePhotos(photos))
      dispatch(closeAddToAlbum())
    } catch (error) {
      console.log(error)
      Alerter.error('Albums.add_photos.error.reference')
    }
  },
  fetchAlbums: () => dispatch(fetchAlbums())
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(AddToAlbumModal))
