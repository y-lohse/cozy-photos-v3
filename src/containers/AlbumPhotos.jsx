import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import styles from '../styles/layout'

import { AlbumToolbar, getAlbum, getAlbumPhotos, fetchAlbum, fetchAlbumPhotos, updateAlbum } from '../ducks/albums'

import BoardView from './BoardView'
import Topbar from '../components/Topbar'
import Alerter from '../components/Alerter'

export class AlbumPhotos extends Component {
  constructor (props) {
    super(props)
    this.state = {
      editing: false
    }
  }

  componentWillMount () {
    this.props.fetchAlbum(this.props.router.params.albumId)
  }

  editAlbumName () {
    this.setState({ editing: true })
  }

  renameAlbum (name) {
    if (name.trim() === '') {
      Alerter.error('Error.album_rename_abort')
      return
    } else if (name === this.props.album.name) {
      this.setState({ editing: false })
      return
    }

    let updatedAlbum = { ...this.props.album, name: name }
    this.props.updateAlbum(updatedAlbum)
      .then(() => {
        this.setState({ editing: false })
      })
      .catch(() => {
        Alerter.error('Error.generic')
      })
  }

  render () {
    if (!this.props.album) {
      return null
    }
    const { album, photos, fetchMorePhotos } = this.props
    const { editing } = this.state
    return (
      <div className={styles['pho-content-wrapper']}>
        {(album.name && photos) &&
          <Topbar viewName='albumContent' albumName={album.name} editing={editing} onEdit={this.renameAlbum.bind(this)} >
            <AlbumToolbar album={album} photos={photos.entries} onRename={this.editAlbumName.bind(this)} />
          </Topbar>
        }
        {photos &&
          <BoardView
            album={album}
            photoLists={[{ photos: photos.entries }]}
            fetchStatus={photos.fetchStatus}
            hasMore={photos.hasMore}
            photosContext='album'
            onFetchMore={() => fetchMorePhotos(album, photos.entries.length)}
          />
        }
        {this.renderViewer(this.props.children)}
      </div>
    )
  }

  renderViewer (children) {
    if (!children) return null
    return React.Children.map(children, child => React.cloneElement(child, {
      photos: this.props.photos.entries
    }))
  }
}

const mapStateToProps = (state, ownProps) => ({
  album: getAlbum(state, ownProps.router.params.albumId),
  photos: getAlbumPhotos(state, ownProps.router.params.albumId)
})

export const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchAlbum: (id) => dispatch(fetchAlbum(id)),
  fetchMorePhotos: (album, skip) => dispatch(fetchAlbumPhotos(album, skip)),
  updateAlbum: (album) => dispatch(updateAlbum(album))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(AlbumPhotos))
