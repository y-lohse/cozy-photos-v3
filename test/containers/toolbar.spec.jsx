'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { mockT } from 'cozy-ui/react/I18n'
import { Toolbar } from '../../src/containers/Toolbar'

describe('Toolbar component', () => {
  it('should be rendered correctly', () => {
    const component = shallow(
      <Toolbar t={mockT} />
    ).node
    expect(component).toMatchSnapshot()
  })
})
