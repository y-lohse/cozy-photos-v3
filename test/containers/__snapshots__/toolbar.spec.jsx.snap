exports[`Toolbar component should be rendered correctly 1`] = `
<div
  className="pho-toolbar"
  role="toolbar">
  
</div>
`;
